sql/core/src/main/scala/org/apache/spark/sql/execution/columnar/InMemoryTableScanExec.scala

```diff
@@ -98,14 +98,23 @@ case class InMemoryTableScanExec(
     val columnarBatch = new ColumnarBatch(columnVectors.asInstanceOf[Array[ColumnVector]])
     columnarBatch.setNumRows(rowCount)
 
-    for (i <- attributes.indices) {
-      ColumnAccessor.decompress(
-        cachedColumnarBatch.buffers(columnIndices(i)),
-        columnarBatch.column(i).asInstanceOf[WritableColumnVector],
-        columnarBatchSchema.fields(i).dataType, rowCount)
+    // Decompress cached batch if the task is not completed, or there is no task context.
+    if (taskContext.isEmpty || !taskContext.get.isCompleted) {
+      for (i <- attributes.indices) {
+        ColumnAccessor.decompress(
+          cachedColumnarBatch.buffers(columnIndices(i)),
+          columnarBatch.column(i).asInstanceOf[WritableColumnVector],
+          columnarBatchSchema.fields(i).dataType, rowCount)
+      }
+    }
+    // When the task is already completed, it is not valid to access a closed batch.
+    if (taskContext.nonEmpty && taskContext.get.isCompleted) {
+      columnarBatch.close()
```

Comment:
So, does this for the case when `addTaskCompletionListener` is not called yet?

Reply:
Yea.